Avoid unnecessary re-computation autograd codegen 21s -> 15s

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59852 pre compute regex and match simple signature autograd codegen 15s -> 12s
* #59850 refactor yaml loader import, no runtime change
* #59849 Use C dumper if possible aten codegen 23s -> 13s
* #59848 avoid error string formatting aten codegen 28s -> 23s
* **#59847 Avoid unnecessary re-computation autograd codegen 21s -> 15s**

This whole stack does not change anything to the codegened code

Differential Revision: [D29063817](https://our.internmc.facebook.com/intern/diff/D29063817)

albanD committed:
Avoid unnecessary re-computation autograd codegen 21s -> 15s

[ghstack-poisoned]

albanD committed:
Update on "Avoid unnecessary re-computation autograd codegen 21s -> 15s"


This whole stack does not change anything to the codegened code


[ghstack-poisoned]

soulitzer reviewed (approved):
Nice speed up!

For the caching though maybe we can use `@functools.cache` decorator instead?

albanD committed:
Update on "Avoid unnecessary re-computation autograd codegen 21s -> 15s"


This whole stack does not change anything to the codegened code


[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29063817).
